### PR TITLE
[Feat](nereids)Support date_trunc function in partition prune

### PR DESCRIPTION
**Pull request by feiniaofeiafei**
The nereids range partition pruner does not support date_trunc(dt, 'day')>10. This pr supports to compute date_trunc(dt, 'day')>10. For example, there is a partition ('2020-01-01 10:00:00', '2020-02-05 10:00:00') ,partition column is dt,  and a predicate date_trunc(dt, 'day')<'2019-01-01 ', this partition can be pruned.

### CODE REVIEW

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DateTrunc.java [hunk 1]
```diff
@@ -91,4 +92,19 @@ public List<FunctionSignature> getSignatures() {
...
+    }
+
+    @Override
+    public int getMonotonicFunctionChildIndex() {
+        return 0;
+    }
+
+    @Override
+    public Monotonicity getMonotonicity() {
+        return new Monotonicity(true, true);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DateTrunc.java:108]: maybe, u should use some static variables generated in Monotonicity class. refer to `org.apache.doris.nereids.properties.PhysicalProperties`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Function.java [hunk 2]
```diff
@@ -46,4 +58,16 @@ public boolean isHighOrder() {
...
+        return this;
+    }
+
+    public int getMonotonicFunctionChildIndex() {
+        return -1;
+    }
+
+    public Monotonicity getMonotonicity() {
+        return new Monotonicity(false, true);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Function.java:72]: add comment to these interfaces

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Function.java [hunk 3]
```diff
@@ -46,4 +58,16 @@ public boolean isHighOrder() {
     public final String getName() {
         return name;
     }
+
+    public Expression withConstantArgs(Literal literal) {
+        return this;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Function.java:63]: return this is not a good idea. i think we should throw exceptions if on function not support Monotonicity

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Function.java [hunk 4]
```diff
@@ -27,6 +28,17 @@
...
+    /**Monotonicity*/
+    public class Monotonicity {
+        public boolean isMonotonic;
+        public boolean isPositive;
+
+        public Monotonicity(boolean isMonotonic, boolean isPositive) {
+            this.isMonotonic = isMonotonic;
+            this.isPositive = isPositive;
+        }
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Function.java:40]: move out from function, it should be expression trait?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OneRangePartitionEvaluator.java [hunk 5]
```diff
@@ -616,6 +630,19 @@ private List<Literal> toMultiNereidsLiterals(PartitionKey partitionKey) {
...
+        EvaluateRangeResult result = super.visitDateTrunc(dateTrunc, context);
+        if (!(result.result instanceof DateTrunc)) {
+            return result;
+        }
+        Expression dateTruncChild = dateTrunc.child(0);
+        if (partitionSlotContainsNull.containsKey(dateTruncChild)) {
+            partitionSlotContainsNull.put(dateTrunc, true);
+        }
+        return computeMonotonicFunctionRange(result);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OneRangePartitionEvaluator.java:644]: if we add a new trait class Monotonic for expression, could we here use visit Monotoic to do same things for all monotonic functions? then we only need add trait on different function

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/ColumnRange.java [hunk 6]
```diff
@@ -69,6 +69,10 @@ public boolean isEmptyRange() {
         return rangeSet.isEmpty();
     }
 
+    public boolean isCompletelyInfinite() {
+        return !span().hasLowerBound() && !span().hasUpperBound();
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/ColumnRange.java:73]: I think this is not completeInfinite
```java
new ColumnRange(
    ColumnBound.lessThen(Literal.of(0)))
        .union(ColumnRange.greaterThan(Literal.of(10))
);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OneRangePartitionEvaluator.java [hunk 7]
```diff
@@ -194,22 +197,14 @@ public EvaluateRangeResult visit(Expression expr, EvaluateRangeInput context) {
...
-
     @Override
     public EvaluateRangeResult visitSlot(Slot slot, EvaluateRangeInput context) {
         // try to replace partition slot to literal
         PartitionSlotInput slotResult = context.slotToInput.get(slot);
-        return slotResult == null
-                ? new EvaluateRangeResult(slot, context.defaultColumnRanges, ImmutableList.of())
-                : new EvaluateRangeResult(slotResult.result, slotResult.columnRanges, ImmutableList.of());
+        assert slotResult != null;
+        assert slotResult.columnRanges.containsKey(slot);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OneRangePartitionEvaluator.java:205]: assert only throw exception in test environment

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OneRangePartitionEvaluator.java (hunk 5 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OneRangePartitionEvaluator.java:644]: we should use trait interface
```java
public interface Monotonic extends ExpressionTrait {
  boolean isPositive(); 
}

public class DateTrunc implement Monotonic {
    @Override
    public isPositive() {
        return true;
    }
}
```

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run cloud_p0

**Comment by feiniaofeiafei** [comment]
run buildall

**Review by 924060929 — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**